src/kernel/lthread/lthread.h

```diff
@@ -0,0 +1,130 @@
+ *      // following pattern.
+ *      goto lthread_resume(self, &&start);
+ *
+ *   start:
+ *        // If example() runs for the first time, goto jumps here.
+ *        ...
+ *
+ *   mutex_retry:
+ *       if (mutex_trylock_lthread(self, &mutex) == -EAGAIN) {
+ *           // mutex is held by another schedee, exit from function in order
+ *           // to be woken up later. The next time execution will start with
+ *           // goto mutex_retry.
+ *           return lthread_yield(&&start, &&mutex_retry);
+ *       }
+ *
+ *       ... // executes with mutex held
+ *
+ *       mutex_unlock_lthread(self, &mutex);
+ *
+ *       ...
+ *
+ *       // Return value 0 corresponds to the start label.
+ *       return 0;
+ *  }
+ * @endcode
+ *
+ * @date 10.12.2013
+ * @author Andrey Kokorev
+ *          - Initial light thread prototype
+ * @author Vita Loginova
+ *          - Rewriting from scratch
+ */
+
+#ifndef _KERNEL_LTHREAD_H_
+#define _KERNEL_LTHREAD_H_
+
+#include <kernel/sched.h>
+#include <kernel/lthread/lthread_sched_wait.h>
+
+
+struct lthread {
+	struct schedee schedee;
+	int (*run)(struct lthread *); /**< Start routine */
+	int label_offset;             /**< Used for calculating label for goto */
+	/* Auxiliary structure for waiting with timeout. */
+	struct sched_wait_info info;
+	struct schedee *joining; /**< joining schedee with stack */
+};
+
+/**
+ * Initializes a new light thread.
+ * @param lt
+ *   The light thread to initialize.
+ * @param run
+ *   The light thread start routine.
+ */
+extern void lthread_init(struct lthread *lt, int (*run)(struct lthread *));
+
+/**
+ * Waits for @p lt is not waiting or scheduled and after that resets @p lt
+ * into initial state. @p lt can be safely launched after executing this
+ * function.
+ * @note
+ *   Can be used only by schedee with stack.
+ * @note
+ *   It is the user's responsibility to prevent lthread waking up in its run
+ *   function.
+ * @param lt
+ *   The light thread to join.
+ */
+extern void lthread_join(struct lthread *lt);
+
+/**
+ * Wakes @p lt up or launches it for the first time.
+ * @param lt
+ *   The light thread to launch.
+ */
+extern void lthread_launch(struct lthread *lt);
+
+/**
+ * Calculates a label offset in order to call #lthread_resume() with
+ * @p target_label.
+ * @note
+ *   This function should be used in pair with return statement.
+ * @param initial_label
+ *   The beginning label.
+ * @param target_label
+ *   Label from which lthread should be resumed.
+ */
+#define lthread_yield(initial_label, target_label) \
+			(target_label) - (initial_label);
+
+/**
+ * Returns label from which run function should start with.
+ * @note
+ *   This function should be used in pair with goto statement.
+ * @param lt
+ *   The current light thread.
+ * @param initial_label
+ *   The beginning label.
+ */
+#define lthread_resume(lt, initial_label) \
+			*((lt)->label_offset + (initial_label))
```

Comment:
Why are these macros at the first place? Inline functions would be better here, IMO.